rcmgr.go

```diff
@@ -0,0 +1,448 @@
+package rcmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+type ResourceManager struct {
+	limits Limiter
+
+	system    *SystemScope
+	transient *TransientScope
+
+	mx    sync.Mutex
+	svc   map[string]*ServiceScope
+	proto map[protocol.ID]*ProtocolScope
+	peer  map[peer.ID]*PeerScope
+
+	cancelCtx context.Context
+	cancel    func()
+	wg        sync.WaitGroup
+}
+
+var _ network.ResourceManager = (*ResourceManager)(nil)
+
+type SystemScope struct {
+	*ResourceScope
```

Comment:
yes it does! We pass it around to the scopes internally, for the constraints.

Reply:
Following sync discussion, this pointer would be interior and it could be passed to objects that can outlive the outer object (in the case of transactions), so we'll keep the pointer.